macros/src/query.rs

```diff
@@ -193,6 +203,327 @@ pub fn derive(input: DeriveInput) -> Result<TokenStream2> {
     })
 }
 
+fn derive_enum(enum_ident: Ident, vis: Visibility, data: DataEnum, lifetime: Lifetime) -> Result<TokenStream2> {
+    let mut dangling_variant = None;
+    let mut fetch_variants = TokenStream2::new();
+    let mut state_variants = TokenStream2::new();
+    let mut query_get_variants = TokenStream2::new();
+    let mut fetch_access_variants = TokenStream2::new();
+    let mut fetch_borrow_variants = TokenStream2::new();
+    let mut fetch_prepare_variants = TokenStream2::new();
+    let mut fetch_execute_variants = TokenStream2::new();
+    let mut fetch_release_variants = TokenStream2::new();
+    let mut fetch_for_each_borrow = TokenStream2::new();
+
+    for variant in &data.variants {
+        let (fields, queries): (Vec<syn::Member>, Vec<TokenStream2>) = match variant.fields {
+            syn::Fields::Named(ref fields) => fields
+                .named
+                .iter()
+                .map(|f| {
+                    (
+                        syn::Member::Named(f.ident.clone().unwrap()),
+                        query_ty(&lifetime, &f.ty),
+                    )
+                })
+                .unzip(),
+            syn::Fields::Unnamed(ref fields) => fields
+                .unnamed
+                .iter()
+                .enumerate()
+                .map(|(i, f)| {
+                    (
+                        syn::Member::Unnamed(syn::Index {
+                            index: i as u32,
+                            span: Span::call_site(),
+                        }),
+                        query_ty(&lifetime, &f.ty),
+                    )
+                })
+                .unzip(),
+            syn::Fields::Unit => continue,
+        };
+
+        let ident = variant.ident.clone();
+
+        if fields.is_empty() {
+            // Ignore empty variants since otherwise they would always match
+            continue;
+        }
+
+        let fetches = queries
+            .iter()
+            .map(|ty| quote! { <#ty as ::hecs::Query>::Fetch })
+            .collect::<Vec<_>>();
+
+        if dangling_variant.is_none() {
```

Comment:
Will do, however the change won't make empty enums work since the Derive macro checks for a lifetime and (afaik) having a lifetime on an empty enum is a compile error from Rust itself

Reply:
[fc0236c](https://github.com/Ralith/hecs/pull/384/commits/fc0236cfafe5e0427ab39c4cc543156d89bfc6e2)